google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,249 @@
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_months months
+            Interval.new months, 0, 0
+          end
+
+          # Returns an Interval instance with the given days.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_days days
+            Interval.new 0, days, 0
+          end
+
+          # Returns an Interval instance with the given seconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_seconds seconds
+            nanoseconds = seconds_to_nanoseconds seconds
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given milliseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_milliseconds milliseconds
+            nanoseconds = milliseconds * NANOSECONDS_IN_A_MILLISECOND
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given microseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_microseconds microseconds
+            nanoseconds = microseconds * NANOSECONDS_IN_A_MICROSECOND
+            Interval.new 0, 0, nanoseconds
+          end
+
+          # Returns an Interval instance with the given nanoseconds.
+          #
+          # @param [Integer]
+          # @return [Interval]
+          def from_nanoseconds nanoseconds
+            Interval.new 0, 0, nanoseconds
+          end
+        end
+
+        def to_s
+          # Memoizing it as the logic can be a bit heavy
+          @string_representation ||= self.to_string
+        end
+
+        private
+
+        def initialize months, days, nanoseconds
+          if (months > MAX_MONTHS || months < MIN_MONTHS)
+            raise ArgumentError, "The Interval class supports months from #{MIN_MONTHS} to #{MAX_MONTHS}."
+          end
+          @months = months
+
+          if (days > MAX_DAYS || days < MIN_DAYS)
+            raise ArgumentError, "The Interval class supports days from #{MIN_DAYS} to #{MAX_DAYS}."
+          end
+          @days = days
+
+          if (nanoseconds > MAX_NANOSECONDS || nanoseconds < MIN_NANOSECONDS)
+            raise ArgumentError, "The Interval class supports nanoseconds from #{MIN_NANOSECONDS} to #{MAX_NANOSECONDS}"
+          end
+          @nanoseconds = nanoseconds
+        end
+
+        def flip_if_needed value
+          if value < 0
+            return -1
+          end
+
+          return 1
+        end
+
+        def to_string
+          years = 0
+          months = 0
+          days = @days
+          hours = 0
+          minutes = 0
+          seconds = 0
+          remaining_nanoseconds = @nanoseconds
+
+          years = @months.fdiv(12).truncate
+          months = @months % (flip_if_needed(@months) * 12)
+          hours = Integer(remaining_nanoseconds / 3_600_000_000_000)
```

Comment:
Why are you using constants for some things (above) but not here (especially because you've already defined `NANOSECONDS_IN_AN_HOUR`)?